File: api/src/schema.rs
use edgedb_derive::Queryable;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Date {
    pub id: Option<Uuid>,
    pub day: Option<i16>,
    pub month: Option<i16>,
    pub year: Option<i32>,
    pub display: Option<String>,
    pub births: Option<Vec<Person>>,
    pub deaths: Option<Vec<Person>>,
    pub artist_starts: Option<Vec<Artist>>,
    pub artist_ends: Option<Vec<Artist>>,
    pub compositions: Option<Vec<Composition>>,
    pub arrangements: Option<Vec<Composition>>,
    pub tracks_recorded: Option<Vec<Track>>,
    pub tracks_released: Option<Vec<Track>>,
    pub tracks_mastered: Option<Vec<Track>>,
    pub tracks_mixed: Option<Vec<Track>>,
    pub albums_released: Option<Vec<Album>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Person {
    pub id: Option<Uuid>,
    pub first_name: Option<String>,
    pub last_name: Option<String>,
    pub aliases: Option<Vec<String>>,
    pub birth_date: Option<Date>,
    pub death_date: Option<Date>,
    pub is_alive: Option<bool>,
    pub full_name: Option<String>,
    pub age: Option<Vec<i32>>,
    pub compositions: Option<Vec<Composition>>,
    pub arrangements: Option<Vec<Composition>>,
    pub lyrics: Option<Vec<Composition>>,
    pub tracks: Option<Vec<Track>>,
    pub is_composer: Option<bool>,
    pub is_arranger: Option<bool>,
    pub is_lyricist: Option<bool>,
    pub is_player: Option<bool>,
    pub is_producer: Option<bool>,
    pub instruments: Option<Vec<Instrument>>,
    pub groups: Option<Vec<Artist>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub enum NoteName {
    C,
    D,
    E,
    F,
    G,
    A,
    B,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub enum Accidental {
    Flat,
    Natural,
    Sharp,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Note {
    pub name: Option<NoteName>,
    pub accidental: Option<Accidental>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Instrument {
    pub id: Option<Uuid>,
    pub name: Option<String>,
    pub aliases: Option<Vec<String>>,
    pub tuning: Option<Note>,
    pub players: Option<Vec<Player>>,
    pub player_names: Option<Vec<String>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub enum Mode {
    Major,
    Minor,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Key {
    pub root: Option<Note>,
    pub mode: Option<Mode>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub enum Denominator {
    One,
    Two,
    Four,
    Eight,
    Sixteen,
    ThirtyTwo,
    SixtyFour,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct TimeSignature {
    pub numerator: Option<i32>,
    pub denominator: Option<Denominator>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Composition {
    pub id: Option<Uuid>,
    pub title: Option<String>,
    pub composers: Option<Vec<Person>>,
    pub lyricists: Option<Vec<Person>>,
    pub arrangers: Option<Vec<Person>>,
    pub date_composed: Option<Date>,
    pub date_arranged: Option<Date>,
    pub instrumentation: Option<Vec<Instrument>>,
    pub key: Option<Key>,
    pub time_signature: Option<TimeSignature>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Player {
    pub id: Option<Uuid>,
    pub person: Option<Person>,
    pub instrument: Option<Instrument>,
    pub display: Option<String>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Artist {
    pub id: Option<Uuid>,
    pub name: Option<String>,
    pub members: Option<Vec<Person>>,
    pub date_start: Option<Date>,
    pub date_end: Option<Date>,
    pub albums: Option<Vec<Album>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Track {
    pub id: Option<Uuid>,
    pub title: Option<String>,
    pub compositions: Option<Vec<Composition>>,
    pub players: Option<Vec<Player>>,
    pub date_recorded: Option<Date>,
    pub date_released: Option<Date>,
    pub date_mastered: Option<Date>,
    pub date_mixed: Option<Date>,
    pub number: Option<i16>,
    pub duration: Option<String>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Series {
    pub id: Option<Uuid>,
    pub name: Option<String>,
    pub label: Option<Vec<Label>>,
    pub albums: Option<Vec<Album>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Label {
    pub id: Option<Uuid>,
    pub name: Option<String>,
    pub series: Option<Vec<Series>>,
    pub albums: Option<Vec<Album>>,
    pub artists: Option<Vec<Artist>>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Disc {
    pub id: Option<Uuid>,
    pub disc_title: Option<Vec<String>>,
    pub number: Option<i32>,
    pub tracks: Option<Vec<Track>>,
    pub album: Option<Album>,
    pub title: Option<Vec<String>>,
    pub duration: Option<String>,
}

#[derive(Debug, Deserialize, Queryable, Serialize)]
#[edgedb(json)]
pub struct Album {
    pub id: Option<Uuid>,
    pub title: Option<String>,
    pub artists: Option<Vec<Artist>>,
    pub producers: Option<Vec<Person>>,
    pub discs: Option<Vec<Disc>>,
    pub label: Option<Label>,
    pub catalog_number: Option<i32>,
    pub series: Option<Series>,
    pub series_number: Option<i32>,
    pub date_released: Option<Date>,
    pub date_recorded: Option<Date>,
    pub tracks: Option<Vec<Track>>,
    pub disc_total: Option<i32>,
    pub duration: Option<String>,
}

File: ui/src/schema.ts
export interface Date {
    id: string;
    type_name: string;
    display: string;
    day: number;
    month: number;
    year: number;
    births: Person[];
    deaths: Person[];
    artist_starts: Artist[];
    artist_ends: Artist[];
    compositions: Composition[];
    arrangements: Composition[];
    tracks_recorded: Track[];
    tracks_released: Track[];
    tracks_mastered: Track[];
    tracks_mixed: Track[];
    albums_released: Album[];
}

export interface Person {
    id: string;
    type_name: string;
    display: string;
    first_name: string;
    last_name: string;
    aliases: String[];
    birth_date: Date;
    death_date: Date;
    is_alive: boolean;
    full_name: string;
    age: number[];
    compositions: Composition[];
    arrangements: Composition[];
    lyrics: Composition[];
    tracks: Track[];
    is_composer: boolean;
    is_arranger: boolean;
    is_lyricist: boolean;
    is_player: boolean;
    is_producer: boolean;
    instruments: Instrument[];
    artists: Artist[];
}

enum NoteName {
    C = 0,
    D = 1,
    E = 2,
    F = 3,
    G = 4,
    A = 5,
    B = 6,
}

enum Accidental {
    Flat = 0,
    Natural = 1,
    Sharp = 2,
}

export interface Note {
    id: string;
    type_name: string;
    display: string;
    name: NoteName;
    accidental: Accidental;
}

export interface Instrument {
    id: string;
    type_name: string;
    display: string;
    name: string;
    aliases: string[];
    tuning: Note;
    players: Player[];
}

enum Mode {
    Major = 0,
    Minor = 1,
}

export interface Key {
    id: string;
    type_name: string;
    display: string;
    root: Note;
    mode: Mode;
}

enum Denominator {
    One = 0,
    Two = 1,
    Four = 2,
    Eight = 3,
    Sixteen = 4,
    ThirtyTwo = 5,
    SixtyFour = 6,
}

export interface TimeSignature {
    id: string;
    type_name: string;
    display: string;
    numerator: number;
    denominator: Denominator;
}

export interface Composition {
    id: string;
    type_name: string;
    display: string;
    title: string;
    composers: Person[];
    lyricists: Person[];
    arrangers: Person[];
    date_composed: Date;
    date_arranged: Date;
    instrumentation: Instrument[];
    key: Key;
    time_signature: TimeSignature;
}

export interface Player {
    id: string;
    type_name: string;
    display: string;
    person: Person;
    instrument: Instrument;
    display: string;
}

export interface Artist {
    id: string;
    type_name: string;
    display: string;
    name: string;
    members: Person[];
    date_start: Date;
    date_end: Date;
    albums: Album[];
}

export interface Track {
    id: string;
    type_name: string;
    display: string;
    title: string;
    compositions: Composition[];
    players: Player[];
    date_recorded: Date;
    date_released: Date;
    date_mastered: Date;
    date_mixed: Date;
    number: number;
    duration: string;
}

export interface Series {
    id: string;
    type_name: string;
    display: string;
    name: string;
    label: Label[];
    albums: Album[];
}

export interface Label {
    id: string;
    type_name: string;
    display: string;
    name: string;
    series: Series[];
    albums: Album[];
    artists: Artist[];
}

export interface Disc {
    id: string;
    type_name: string;
    display: string;
    disc_title: String[];
    number: number;
    tracks: Track[];
    album: Album;
    title: String[];
    duration: string;
}

export interface Album {
    id: string;
    type_name: string;
    display: string;
    title: string;
    artists: Artist[];
    producers: Person[];
    discs: Disc[];
    label: Label;
    catalog_number: number;
    series: Series;
    series_number: number;
    date_released: Date;
    date_recorded: Date;
    tracks: Track[];
    disc_total: number;
    duration: string;
}

export type DatabaseProps =
    | Date
    | Person
    | Instrument
    | Composition
    | Artist
    | Album
    | Track
    | Label
    | Series;
